### DISCUSSION TITLE
Using Coyote to generate Traces

### DISCUSSION THREAD

**Discussion by Arhaan** [microsoft/coyote#486] [Q&A] [answered]
Hi
I have been trying for a while to use Coyote to generate the trace of an execution of a program. A trace is a log of all the reads and writes happening in the order in which they happened, along with the thread that caused that. Can anyone point me to the files/parts of code I should edit to achieve this?

**Comment by pdeligia** [comment, marked as answer]
@Arhaan, Coyote does not give you automatically such a log as it does not support instrumenting every single read/write location (focuses on high-level race conditions instead, and to scale in large programs). Saying this, we have some experimental instrumentations (most are disabled by default) that check for a subset of race conditions (such as on collections, class field accesses, etc -- see all the `*RaceChecking*` options on https://microsoft.github.io/coyote/#ref/Microsoft.Coyote/Configuration/). But these might not serve the purpose you want, or have the logging you want, nor be complete (hence the experimental).

Instead, one option you have is to do your own binary rewriting (e.g. using Mono.Cecil, seperate from Coyote's binary rewriting) which (1) adds read/write logs at the places you want them, and (2) invokes this [Interleave](https://microsoft.github.io/coyote/#ref/Microsoft.Coyote.Runtime/SchedulingPoint/Interleave/) API from Coyote to ask it to explore an interleaving at places you care about (e.g. before each read/write instruction). You can see many instrumentation examples how to achieve this either on the Mono.Cecil repository, or inside Coyote's [binary rewriting logic](https://github.com/microsoft/coyote/tree/main/Source/Test/Rewriting). You could even build a local version of Coyote adding your own data race checking instrumentation pass alongside the [existing ones](https://github.com/microsoft/coyote/tree/main/Source/Test/Rewriting/Passes/Rewriting) in Coyote, and if it is generic enough, please consider contributing back via a PR as more people in the community might contribute from this!

Then you invoke Coyote to run a systematic test that includes your logs and interleaving exploration points. That will probably get you closest to what you want. Hope this helps.
